diff --git a/lib/feedparser/api.py b/lib/feedparser/api.py
index 614bd2d..12eafd2 100644
--- a/lib/feedparser/api.py
+++ b/lib/feedparser/api.py
@@ -60,6 +60,7 @@ from .sanitizer import replace_doctype
 from .sgml import *
 from .urls import _convert_to_idn, _makeSafeAbsoluteURI
 from .util import FeedParserDict
+from . import USER_AGENT
 
 bytes_ = type(b'')
 unicode_ = type('')
diff --git a/lib/feedparser/util.py b/lib/feedparser/util.py
index f7c02c0..df36b3e 100644
--- a/lib/feedparser/util.py
+++ b/lib/feedparser/util.py
@@ -122,9 +122,23 @@ class FeedParserDict(dict):
 
     def __setitem__(self, key, value):
         key = self.keymap.get(key, key)
-        if isinstance(key, list):
-            key = key[0]
-        return dict.__setitem__(self, key, value)
+        if key == 'newznab_attr':
+            if isinstance(value, dict) and value.keys() == ['name', 'value']:
+                key = value['name']
+                value = value['value']
+
+            if not dict.__contains__(self, 'categories'):
+                dict.__setitem__(self, 'categories', [])
+
+            if key == 'category':
+                self['categories'].append(value)
+            else:
+                dict.__setitem__(self, key, value)
+        else:
+            if isinstance(key, list):
+                key = key[0]
+
+            return dict.__setitem__(self, key, value)
 
     def setdefault(self, key, value):
         if key not in self:
